requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,189 @@
+
+#### Why not?
+
+Copying would mean a need to maintain such underlying links (e.g. a cross-project ID, adjacent to the primary key ID) for all entities we'd like to work across projects, which would be dashboards, tiles, insights, actions and cohorts at the minimum. In one way, this could be nice if you'd like to develop, say, a dashboard in the test environment and only copy it to the production environment when final, while still being able to diverge between projects. The major UX tradeoff would be the act of manual syncing, making the experience cumbersome for users at scale. This would work, but it wouldn't be turnkey. And the whole tree of dependencies would be massive in terms of entities in Postgres. Support reports around inconsistencies with dependency syncing would be hard to avoid.
+
+## Dismissed solution B: Dashboards/insights/etc. sharing between projects
+
+What if instead of copying whole trees of entities we shared them directly? This could be even less work for the user – no need to sync manually – and for us – no need to copy dozens of entities to share a single dashboard between two projects. Also no extra work in terms of permissioning or ingestion.
+
+#### Why not?
+
+The UX would still be a little clunky, with a need to manually set up multiple full projects for each environment. It's the taxonomy that's the killer though – shared taxonomy seems to be a relatively common ask in itself, but it's also key for analysis in an implicit way. Group type indexes in particular MUST match exactly, or insights will show results for completely different data. Property types matter – numbers are treated differently from strings, and then there are also dates. A deeper level of integration is needed for this to work reliably.
+
+## Proposed solution: Environments
+
+We can make environments a first-class entity in PostHog – each project would be subdivided into one or more environments. The *data* would be isolated between projects, but the *taxonomy* and *analysis setup* would be shared (see "Summarizing needs" for what should be isolated and what should be shared exactly).
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition). This would involve no copying though, so there would be no effort needed to sync things.
+
+To support large customers with access control needs, we'd need to add environment-specific project membership to the existing project access control system.
+
+To support the common feature flags workflow of rollout conditions being different between dev and production, we'd make flag rollout groups always environment-specific.
```

Comment:
I think this will apply to anything that depends on flags too then?

For example, if you have a survey in dev with a targeting flag that's promoted to prod, the dev flag should be promoted as well with the same conditions

Reply:
Also, I'm assuming this mean sync remains then, but becomes a lot more dumb? Because if someone needs to _copy_ dev rollout conditions to prod they should be able to, since these are different per environment. I'm _hoping_ we can have a similar flow as 'promote' for rest of the things, rather than a completely different one for flags, that basically updates one release condition to be the other.

Since cohorts/actions are shared, having the same IDs, this should be pretty reasonable to do with no real syncing logic required.

Reply:
Also, assuming we still never want to share anything across teams (projects) right?

Reply:
Maybe rollout groups should _optionally_ be environment-specific then? Not really more complex backend-wise, just trickier UI-wise – but would support _all_ the use cases

Reply:
Added this as part of requirements